server/shared/src/test/scala/org/http4s/server/middleware/ThrottleSuite.scala

```diff
@@ -27,107 +28,124 @@ import org.http4s.Request
+        (checkTokensUpToCapacity, testee.takeToken.map(_.isInstanceOf[TokenUnavailable]))
+          .mapN(_ && _)
+      }.assert
+
+      TestControl.executeEmbed(prog)
+    }
   }
 
   test("LocalTokenBucket should add another token at specified interval when not at capacity") {
-    val ctx = TestContext()
-
-    val capacity = 1
-    val createBucket =
-      TokenBucket.local[IO](capacity, 100.milliseconds)
+    forAllF(genFiniteDuration) { (someRefillTime: FiniteDuration) =>
+      val capacity = 1
+      val createBucket =
+        TokenBucket.local[IO](capacity, someRefillTime)
+
+      val takeTokenAfterRefill = createBucket.flatMap { testee =>
+        for {
+          _ <- testee.takeToken
+          unavailable <- testee.takeToken
+          _ <- IO.sleep(someRefillTime)
+          available <- testee.takeToken
+        } yield unavailable -> available
+      }
 
-    val takeTokenAfterRefill = createBucket.flatMap { testee =>
-      testee.takeToken *> IO.sleep(101.milliseconds) *>
-        testee.takeToken
+      TestControl
+        .executeEmbed(takeTokenAfterRefill)
+        .map { case (a, b) => a.isInstanceOf[TokenUnavailable] && b == TokenAvailable }
+        .assert
     }
-
-    takeTokenAfterRefill
-      .map { result =>
-        ctx.advanceAndTick(101.milliseconds)
-        result
-      }
-      .assertEquals(TokenAvailable)
   }
 
   test("LocalTokenBucket should not add another token at specified interval when at capacity") {
-    val ctx = TestContext()
-    val capacity = 5
-    val createBucket =
-      TokenBucket.local[IO](capacity, 100.milliseconds)
-
-    val takeExtraToken = createBucket.flatMap { testee =>
-      val takeFiveTokens: IO[List[TokenAvailability]] = (1 to 5).toList.traverse { _ =>
-        testee.takeToken
+    forAllF(genFiniteDuration) { (someRefillTime: FiniteDuration) =>
+      val capacity = 5
+      val createBucket =
+        TokenBucket.local[IO](capacity, someRefillTime)
+
+      val takeExtraToken = createBucket.flatMap { testee =>
+        val takeFiveTokens: IO[List[TokenAvailability]] = (1 to 5).toList.traverse { _ =>
+          testee.takeToken
+        }
+        takeFiveTokens >> IO.sleep(300.milliseconds) >> testee.takeToken
       }
-      IO.sleep(300.milliseconds) >> takeFiveTokens >> testee.takeToken
-    }
 
-    takeExtraToken
-      .map { result =>
-        ctx.advanceAndTick(300.milliseconds)
-        result
+      TestControl.execute(takeExtraToken).flatMap { control =>
+        for {
+          _ <- control.results.assertEquals(None)
+          _ <- control.tick
+          _ <- control.results.assertEquals(None)
+          interval <- control.nextInterval
+          _ <- control.advanceAndTick(interval)
+          _ <- control.results
+            .map(_.exists {
+              case Outcome.Succeeded(TokenUnavailable(Some(_))) => true
+              case _ => false
+            })
+            .assert
+        } yield ()
```

Comment:
Sorry to be annoying, I think this is the last one! :)

Reply:
I must be blind :-) I thought they were all gone - will fix

Reply:
Done

Reply:
Will make one more passthrough review a bit later but I reckon things are pretty good now :) Thanks for all your work on this!

Reply:
Sure thing! And actually thanks for being my watchdog in this PR! 😊